meta-ostro/lib/supportedrecipes/supportedrecipes.py

```diff
@@ -0,0 +1,346 @@
+    filename = d.getVar('FILE', True)
+    collection = bb.utils.get_file_layer(filename, d)
+    pv = d.getVar('PV', True)
+    summary = d.getVar('SUMMARY', True) or ''
+    homepage = d.getVar('HOMEPAGE', True) or ''
+    src = d.getVar('SRC_URI', True).split()
+    sources = []
+    for url in src:
+        scheme, netloc, path, query, fragment = urlparse.urlsplit(url)
+        if scheme != 'file':
+            parts = path.split(';')
+            if len(parts) > 1:
+                path = parts[0]
+                params = dict([x.split('=') if '=' in x else (x, '') for x in parts[1:]])
+            else:
+                params = {}
+            name = params.get('name', None)
+            sources.append((name, '%s://%s%s' % (scheme, netloc, path)))
+    dumpfile = d.getVar('SUPPORTED_RECIPES_SOURCES_DIR', True) + '/' + pn + filename
+    bb.utils.mkdirhier(os.path.dirname(dumpfile))
+    with open(dumpfile, 'w') as f:
+        # File intentionally kept small by not writing a header line. Guaranteed to contain SOURCE_FIELDS.
+        writer = csv.writer(f)
+        for idx, val in enumerate(sources):
+            name, url = val
+            if name and not len(sources) == 1:
+                fullname = '%s/%s' % (pn, name)
+            elif idx > 0:
+                fullname = '%s/%d' % (pn, idx)
+            else:
+                fullname = pn
+            writer.writerow((fullname, collection, pv, homepage, url, summary))
+
+def check_build(d, event):
+    supported_recipes, files = load_supported_recipes(d)
+    supported_recipes_check = d.getVar('SUPPORTED_RECIPES_CHECK', True)
+    if not supported_recipes_check:
+        return
+
+    # Always add a trailing $ to ensure a full match.
+    isnative_exception = re.compile('(' + '|'.join(d.getVar('SUPPORTED_RECIPES_NATIVE_RECIPES', True).split()) + ')$')
+    isnative_baseclasses = d.getVar('SUPPORTED_RECIPES_NATIVE_BASECLASSES', True).split()
+    valid = ('note', 'warn', 'error', 'fatal')
+    if supported_recipes_check not in valid:
+        bb.fatal('SUPPORTED_RECIPES_CHECK must be set to one of %s, currently is: %s' % ('/'.join(valid), supported_recipes_check))
+    logger = bb.__dict__[supported_recipes_check]
+
+    # See bitbake/lib/bb/cooker.py buildDependTree() for the content of the depgraph hash.
+    # Basically it mirrors the information dumped by "bitbake -g".
+    depgraph = event._depgraph
+    # import pprint
+    # bb.note('depgraph: %s' % pprint.pformat(depgraph))
+
+    dirname = d.getVar('SUPPORTED_RECIPES_SOURCES_DIR', True)
+    report_sources = d.getVar('SUPPORTED_RECIPES_SOURCES', True)
+
+    unsupported = {}
+    sources = []
+    for pn, pndata in depgraph['pn'].iteritems():
+        # We only care about recipes compiled for the target.
+        # Most native ones can be detected reliably because they inherit native.bbclass,
+        # but some special cases have to be hard-coded.
+        # Image recipes also do not matter.
+        def isnative():
```

Comment:
redefining function on every loop run, doesn't make sense. move to outer scope that function.